man/man1/plough.1
.TH "plough" 1

.SH NAME
plough \- Interface for running multiple related TractoR experiments

.SH SYNOPSIS
.B plough \fR[\fB\-GXD\fR] \fB\-C \fIconfig_file\fR [\fB\-V \fIvariable_name\fR] [\fB\-Q \fIgrid_queue\fR] [\fB\-O \fIgrid_options\fR] [\fB\-P \fIfactor\fR] [\fItractor_flags\fR] \fIscript\fR [\fIarguments\fR]\fB
.br
.B plough \-H

.SH DESCRIPTION
The \fBplough\fR program is a wrapper around \fBtractor\fR(1), which allows a particular script to be run with several different sets of parameters. The loop can be easily parallelised across multiple cores, or using the Oracle (formerly Sun) Grid Engine, if the latter is installed.
.PP
The program's flags are all uppercase, to avoid conflicting with those from \fBtractor\fR, and the two sets of flags can therefore be freely mixed. The \fItractor_flags\fR and/or \fIarguments\fR can include references to the loop variables, using their name preceded by a percentage sign ('%'), which will be replaced by the appropriate value on each loop. In addition, the symbol '%%' will be replaced by the loop number.

.SH OPTIONS
.TP
.B \-H
Display usage information and quit.
.TP
.B \-G
Schedule parallel jobs on the Oracle (formerly Sun) Grid Engine. The \fBqsub\fR(1) executable must be on your \fBPATH\fR.
.TP
.B \-X
Cross-apply the loop variables, taking each combination in turn. The total number of instances will be the product of the lengths of all loop variables, rather than the maximum.
.TP
.B \-D
Produce debugging output.
.TP
.B \-C \fIconfig_file\fB
Specify a configuration file in YAML format (see \fBtractor\fR(1) for details). This flag must be specified at least once, and may be used multiple times.
.TP
.B \-V \fIvariable_name\fB
Specify one or more variables over which the loop will be run (comma-separated). These must be defined in a config file. By default all variables with more than one element will be used.
.TP
.B \-Q \fIgrid_queue\fB
Name the grid engine queue on which jobs should be scheduled. Has no effect if '\-G' is not set.
.TP
.B \-O \fIgrid_options\fB
Specify additional options to pass to \fBqsub\fR(1), for example to request certain resources from the grid engine. The options must be quoted if they contain any spaces. Has no effect if '\-G' is not set.
.TP
.B \-P \fIfactor\fB
Set the parallelisation factor for multicore systems. At most this number of threads will be created to call \fBtractor\fR with each combination of parameters. This option is ignored if '\-G' is set. The default is taken from the \fBTRACTOR_PLOUGH_THREADS\fR environment variable, if it is set; otherwise it is 1 (i\.e\. no parallelisation).

.SH EXAMPLES
A simple use case is a call to a TractoR script across multiple sessions, as in
.TP
.PP
plough \-C config.yaml \-q mean %Session@FA
.PP
which calculates the mean voxel value in each session's fractional anisotropy map. Note that the 'config.yaml' file must assign one or more values to the 'Session' variable, each giving the path to a session directory. A more complex example might be
.TP
.PP
plough \-C config.yaml \-V Session \-V TractName \-X \-w %TractName pnt-data %Session DatasetName:%TractName_data SearchWidth:7
.PP
This example runs the 'pnt-data' script, using every possible combination of 'Session' and 'TractName' in turn ('\-X'). Experiments for each 'TractName' will be run within a corresponding subdirectory ('\-w %TractName'). The 'SearchWidth' variable will be common to all instances.

.SH DIAGNOSTICS
Return value is 0 if the child jobs were successfully launched or run, and 1 otherwise. The program will wait for its children unless they are run through the grid engine.

.SH AUTHOR
Jon Clayden <code@clayden.org>


multicore/NEWS
 NEWS/ChangeLog for multicore
------------------------------

0.1-3	2009-02-02
    o	simplify the internal management of child processes and
	consequently remove bugs that lead to poor feeding of cores
	in mclapply() when no precheduling was used


0.1-2	2009-01-09
    o	added mc.preschedule parameter to mclappy() which (if FALSE)
	allows on-demand distribution of FUN calls across cores.

    o	added "silent" parmeter to parallel() and mclapply()
	suppressing output on stdout in child processes

    o	added internal functions closeStdout(), closeStrerr(),
	closeFD(), closeAll(), isChild(), childrenDescriptors() and
	masterDescriptor()

    o	selectChildren() implicitly checks for and removes zombies

    o	removed spurious debugging output in mclapply()

    o	fixed cases in which mclapply() would not preserve names

    o	fixed child list management

    o	in order to make sure that children cannot die before the
	master can collect results, children will specifically wait
	for the master to allow them to exit (SIGUSR1 is used for this
	purpose, so don't use it directly unless you want a child to
	exit even if the master didn't release it).


0.1-1	2009-01-03
    o	added name parameter to parallel(), accordingly collect()
	names the results if the job name is present

    o	bug fix: collect() returned early if no jobs responded for a
	while

    o	bug fix: internal list of children could get corrupted when a
	child was removed (all preceding children were removed as well)

    o	added better debugging support. Set PKG_CFLAGS=-DMC_DEBUG to
	enable debugging output on stdout (it is purposefully not sent
	to the R console to prevent output processing issues in child
	processes)


0.1-0	2009-01-02
    o	initial release


tractor.utils/R/plough.R
ploughExperiment <- function (scriptName, configFiles, variables, tractorFlags, tractorOptions, useGridEngine, crossApply, queueName, qsubOptions, parallelisationFactor, debug)
{
    crossApply <- isTRUE(crossApply == 1)
    useGridEngine <- isTRUE(useGridEngine == 1)
    debug <- isTRUE(debug == 1)
    setOutputLevel(ifelse(debug, OL$Debug, OL$Info))
    
    config <- readYaml(configFiles)
    variableLengths <- sapply(config, length)
    
    variables <- splitAndConvertString(variables, ",", fixed=TRUE)
    if (length(variables) == 0 || all(variables == ""))
        variables <- names(config)[variableLengths > 1]
    else if (!all(variables %in% names(config)))
        report(OL$Error, "Specified variable(s) #{implode(variables[!(variables %in% names(config))],sep=', ',finalSep=' and ')} are not mentioned in the config files")
    
    usingParallel <- FALSE
    if (isValidAs(parallelisationFactor,"integer") && as.integer(parallelisationFactor) > 1)
    {
        if (useGridEngine)
            report(OL$Warning, "Parallelisation factor will be ignored when using the grid engine")
        else if (system.file(package="parallel") != "")
        {
            library(parallel)
            options(mc.cores=as.integer(parallelisationFactor))
            usingParallel <- TRUE
        }
        else
            report(OL$Warning, "The \"parallel\" package is not installed - code will not be parallelised")
    }
    
    if (crossApply)
    {
        n <- prod(variableLengths[variables])
        data <- as.data.frame(expand.grid(config[variables], stringsAsFactors=FALSE), stringsAsFactors=FALSE)
    }
    else
    {
        n <- max(variableLengths[variables])
        data <- as.data.frame(lapply(config[variables], rep, length.out=n), stringsAsFactors=FALSE)
    }
    
    report(OL$Info, "Scheduling #{n} jobs")
    
    tractorPath <- file.path(Sys.getenv("TRACTOR_HOME"), "bin", "tractor")
    
    buildArgs <- function (i)
    {
        currentFlags <- ore.subst("(?<!\\\\)\\%(\\w+)", function(match) data[i,groups(match)], tractorFlags, all=TRUE)
        currentFlags <- ore.subst("(?<!\\\\)\\%\\%", as.character(i), currentFlags, all=TRUE)
        currentOptions <- ore.subst("(?<!\\\\)\\%(\\w+)", function(match) data[i,groups(match)], tractorOptions, all=TRUE)
        currentOptions <- ore.subst("(?<!\\\\)\\%\\%", as.character(i), currentOptions, all=TRUE)
        return (es("#{currentFlags} #{scriptName} #{currentOptions}"))
    }
    
    if (useGridEngine)
    {
        for (i in seq_len(n))
        {
            tempDir <- expandFileName("sgetmp")
            if (file.exists(tempDir))
                unlink(tempDir, recursive=TRUE)
            dir.create(file.path(tempDir,"log"), recursive=TRUE)
            
            currentFile <- file.path(tempDir, es("config.#{i}.yaml"))
            writeYaml(as.list(data[i,,drop=FALSE]), currentFile, capitaliseLabels=FALSE)
            
            qsubScriptFile <- file.path(tempDir, "script")
            qsubScript <- c("#!/bin/sh",
                            "#$ -S /bin/bash",
                            es("TRACTOR_PLOUGH_ID=#{i}"),
                            es("#{tractorPath} -c #{currentFile} #{buildArgs(i)}"))
            writeLines(qsubScript, qsubScriptFile)
            execute("chmod", es("+x #{qsubScriptFile}"))
            
            queueOption <- ifelse(queueName=="", "", es("-q #{queueName}"))
            qsubArgs <- es("-terse -V -wd #{path.expand(getwd())} #{queueOption} -N #{scriptName} -o #{file.path(tempDir,'log')} -e /dev/null -t 1-#{n} #{qsubOptions} #{qsubScriptFile}")
            result <- execute("qsub", qsubArgs, intern=TRUE)
            jobNumber <- as.numeric(ore.match("^(\\d+)\\.?.*$", result)[,1])
            jobNumber <- jobNumber[!is.na(jobNumber)]
            report(OL$Verbose, "Job number for index #{i} is #{jobNumber}")
        }
    }
    else
    {
        parallelApply(seq_len(n), function(i) {
            currentFile <- threadSafeTempFile()
            writeYaml(as.list(data[i,,drop=FALSE]), currentFile, capitaliseLabels=FALSE)
            
            Sys.setenv(TRACTOR_PLOUGH_ID=i)
            execute(tractorPath, es("-c #{currentFile} #{buildArgs(i)}"))
            
            unlink(currentFile)
        })
    }
    
    invisible(NULL)
}
